src/Shared/IEnumerableExtensions.cs

```diff
@@ -0,0 +1,237 @@
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Microsoft.Build.Collections
+{
+    /// <summary>
+    /// A set of extension methods for working with immutable dictionaries.
+    /// </summary>
+    internal static class IEnumerableExtensions
+    {
+        /// <summary>
+        /// Avoids allocating an enumerator when enumerating an <see cref="IEnumerable{T}"/> in many cases.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The <see langword="foreach"/> statement enumerates types that implement <see cref="IEnumerable{T}"/>,
+        /// but will also enumerate any type that has the required methods. Several collection types take advantage of this
+        /// to avoid allocating an enumerator on the heap when used with <see langword="foreach"/> by returning a
+        /// <see langword="struct"/> enumerator. This is in contrast to the interface-based enumerator
+        /// <see cref="IEnumerator{T}"/> which will always be allocated on the heap.
+        /// </para>
+        /// <para>
+        /// This extension method attempts to create a non-allocating struct enumerator to enumerate
+        /// <paramref name="collection"/>. It checks the concrete type of the collection and provides a
+        /// non-allocating path in several cases.
+        /// </para>
+        /// <para>
+        /// Types that can be enumerated without allocation are:
+        /// </para>
+        /// <list type="bullet">
+        ///     <item><description><see cref="IList{T}"/> (and by extension <see cref="List{T}"/> and other popular implementations)</description></item>
+        ///     <item><description><see cref="LinkedList{T}"/></description></item>
+        ///     <item><description><see cref="ImmutableHashSet{T}"/></description></item>
+        ///     <item><description><see cref="ImmutableList{T}"/></description></item>
+        ///     <item><description><see cref="ICollection{T}"/> or <see cref="IReadOnlyCollection{T}"/> having zero count</description></item>
+        /// </list>
+        /// <para>
+        /// If <paramref name="collection"/> is not one of the supported types, the returned enumerator falls back to the
+        /// interface-based enumerator, which will heap allocate. Benchmarking shows the overhead in such cases is low enough
+        /// to be within the measurement error.
+        /// </para>
+        /// </remarks>
+        /// <example>
+        /// <code>
+        /// <![CDATA[IEnumerable<string> collection = ...;
+        ///
+        /// foreach (string item in collection.GetStructEnumerable())
+        /// {
+        ///     // ...
+        /// }]]>
+        /// </code>
+        /// </example>
+        /// <typeparam name="T">The item type that is enumerated.</typeparam>
+        /// <param name="collection">The collections that will be enumerated.</param>
+        /// <returns>The enumerator for the collection.</returns>
```

Comment:
If `foreach` can be optimised this way - shouldn't it be something that should be part of compiler/runtime? Are we possible missing on any already existing opportunity there? @stephentoub?

Or is this bringing benefit only on NetFx? (in which case we migh possibly want to conditionaly compile it only for NetFx?)

Reply:
This appears to be trying to manually special-case a few specific collection types by maintaining a discriminated union of all of their strongly-typed struct enumerators. I would not be surprised if this actually makes things more expensive in various situations, especially on .NET Core where dynamic PGO (on by default as of .NET 8) will already do such special-casing and handle devirtualizing and inlining the enumerator dispatches. In .NET 10 the enumerator allocation itself will also frequently be avoided thanks to stack allocation / escape analysis work currently being done by @AndyAyersMS.

Reply:
The intent with this helper is to selectively apply it in situations where the allocations are significant, so I'd be hesitant to unconditionally apply it for reasons that @stephentoub mentions. Anecdotally, I've applied this change in a couple of places and reverted the change because I saw an increase in CPU cost that wasn't offset by the reduction in allocations.

Reply:
Gotcha.
Sounds helpful on NetFX. We should probably measure impact on `dotnet build` (core) though

Reply:
> handle devirtualizing and inlining the enumerator dispatches

Just to reinterate @Erarndt here, MSBuild uses workstation GC, and is often blocked from burning CPU because its stuck behind a single threaded GC performing a Gen0/Gen1 cleaning up these enumerators. While long term we want Server GC for this process which clean up these allocations with ease, it often better within MSBuild burn additional CPU walking these collections across many threads to avoid the allocation, then it is to be block all threads behind whatever thread triggered the workstation GC.